.github/workflows/main.yml

```diff
@@ -22,22 +36,16 @@ jobs:
         uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
+          cache: pip
 
       - name: Install project dependencies
         run: |
-          pip3 install --upgrade pip
-          pip3 install codecov tox tox-gh-actions
+          pip install --upgrade pip
+          pip install codecov .[tests]
+          pip install '${{ matrix.urllib3-version }}'
```

Comment:
- The first of these two always installs urllib3 2.x without apparent need to. Let's combine these two lines into one?
- Let's append `pip check` at the end to avoid unknown dependency issues
- (Let's add `set -x` to the first line to help CI logs make sense. It's hard to see where each command ends right now.)